### DISCUSSION TITLE
Better support for url typing in create_engine()

### DISCUSSION THREAD

**Discussion by mbeijen** [sqlalchemy/sqlalchemy#12169] [Usage Questions]
pydantic has nice specific types for PostgreSQL, MariaDB and MySQL DSNs. However if we have such a type and feed it to SQLAlchemy's `create_engine()` it fails because it's not of type `str` however it can be stringified:

```
>>> import sqlalchemy
>>> import pydantic
>>> dsn = pydantic.parse_obj_as(pydantic.MySQLDsn, "mysql+mysqldb://scott:tiger@localhost/test") 
>>> dsn
MySQLDsn('mysql+mysqldb://scott:tiger@localhost:3306/test')
>>> str(dsn)
'mysql+mysqldb://scott:tiger@localhost:3306/test'
>>> sqlalchemy.create_engine(dsn)
sqlalchemy.exc.ArgumentError: Expected string or URL object, got MySQLDsn('mysql+mysqldb://scott:tiger@localhost:3306/test')
```

I understand I can call `create_engine(str(dsn))` and it works. I also understand it's not desirable to expand SQLAlchemy to allow the pydantic.MySQLDsn and such as parameters for `create_engine()`

But would it be possible to also allow objects that can be stringified as inputs, and stringify them if needed? That would allow me passing these specific types to `create_engine()`

**Comment by CaselIT** [comment]
Hi,

> But would it be possible to also allow objects that can be stringified as inputs, and stringify them if needed? That would allow me passing these specific types to create_engine()

any object can be stringified by calling `str(object)`, so this is equivalent of saying that we accept any and hope for the best. I'm -1 on this approach

If these classes of pydantic are meant to represent strings then they should subclass str.
I'm not really sure what sqlalchemy can do here

**Reply by mbeijen** [reply]
> any object can be stringified by calling `str(object)`, so this is equivalent of saying that we accept any and hope for the best. I'm -1 on this approach

We're not simply hoping for the best: create_engine() feeds the urls into `_parse_url()` which does a good job of making sure the dsn is well-formed. So if I would feed some other object that can be stringified, I'd still get a proper error message.

So the downside would be that we lose typie hints that would state that it needs to be a str or an url object (which is not overly specific or helpful per se), but the upside is that we can provide pydantics specific DSNs (or any other popular typing libraries specific DSN objects, might they exist) to create_engine and it will "Just Work"

**Reply by CaselIT** [reply]
we already have a way of making things just work: if an object is meant to represent a string then make it a string subclass. So pydantic DSNs should subclass it

Overall I'm -1 since you will have wired errors from `_parse_url`.

**Comment by zzzeek** [comment]
Why do you need a `pydantic.MySQLDsn` ?  Are these explicitly compatible and syntax checked for SQLAlchemy dialects?  if so, why wouldn't they subclass our own `URL`, and if not, `str()` would not be an appropriate conversion utility since SQLAlchemy URLs are specific to SQLAlchemy in several ways

**Reply by mbeijen** [reply]
they are indeed syntax checked. below is a really simple example.

```
>>> import pydantic, sqlalchemy
>>> dsn = pydantic.parse_obj_as(pydantic.MySQLDsn, "bysql://scott:tiger@localhost/test") 
pydantic_core._pydantic_core.ValidationError: 1 validation error for function-wrap[wrap_val()]
  URL scheme should be 'mysql', 'mysql+mysqlconnector', 'mysql+aiomysql', 'mysql+asyncmy', 'mysql+mysqldb', 'mysql+pymysql', 'mysql+cymysql' or 'mysql+pyodbc' [type=url_scheme, input_value='fysql+mysqldb://scott:tiger@localhost/test', input_type=str]
```

It's interesting that you state that SQLAlchemy's URLs are specific to SQLAlchemy; indeed I assumed that pydantic's DSNs are specifically for SQLAlchemy but now I'm not so sure ;-)

**Reply by CaselIT** [reply]
that error seems wired. why It's complaining?
`mysql+mysqldb` is in the list of schemes as the 5th value: `'mysql+mysqldb'`

**Reply by mbeijen** [reply]
sorry, that was a cut & paste issue on my side ;-)

**Reply by CaselIT** [reply]
that seem to validate sqlalchemy url. In my opinion it's on them to make it work with sqlalchemy.

Note that I would not be opposed at adding support for a magic method that sqlalchemy checks for, something like `def __sqlalchemy_url__(self) -> URL | str: ...` that an object can implement to indicate that it can provide a valid sqlalchemy url

**Reply by zzzeek** [reply]
our URL format is not any kind of standard, so pydantic has to be importing our parser for this to be maintainable on their end.   In particular things like @ signs in passwords have to be URL escaped and this was not the process previously.      not really sure how pydantic sees this feature if it is indeed importing SQLAlchemy anyway

**Reply by mbeijen** [reply]
Thanks, I've opened a discussion at pydantic now --> https://github.com/pydantic/pydantic/discussions/11074
